File: 935_div3/f.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

using ll = int64_t;

void solve() {
  int n; cin >> n;

  vector<ll> v(n + 1);
  for (int i = 1; i <= n; ++i) cin >> v[i];

  vector<ll> p(n + 1);
  for (int i = 1; i <= n; ++i) cin >> p[i];

  vector<int> indexes(n + 1);
  for (int i = 0; i <=n; ++i) indexes[i] = i;

  // Сортируем indexes по убыванию значений v. Будем срывать грибы в порядке убывания волшебности.
  // Так мы всегда будем последним срывать гриб, который окажется минимальным в корзинке.
  sort(indexes.begin(), indexes.end(), [&](int i, int j) { return v[i] > v[j]; });

  // Вы корзине (bucket) храним индексы грибов, которые нарвали для эликсира.
  unordered_set <int> bucket;
  
  // Номер следующего срываемого гриба.
  int pick_index = 0;

  // Максимальная сила эликсира.
  ll max_power = 0;

  // Наименьшее количество грибов, которые дают эликсир с силой max_power.
  int min_shrooms = n;

  // take -- количество грибов, которые мы хотим сорвать для эликсира. Будем искать минимальное
  // количество грибов с максимальной силой.
  for (int take = 1; take <= n; ++take) {
    // Нужно ли удалить еще один гриб из кандидатов в эликсир.
    if (take >= 2) {
      const int remove = p[take - 1];
      v[remove] = 0;
      bucket.erase(remove);
    }

    // Пока в корзинке меньше грибов, чем take и под дубом еще есть грибы.
    while (bucket.size() < take && pick_index <= n) {
      // Индекс срываемого гриба.
      const int index = indexes[pick_index];
      
      // Если гриб хороший (с еще не зануленной волшебностью).
      if (v[index] > 0) bucket.insert(index);

      // Если получился эликсир лучше, чем раньше.
      if (bucket.size() * v[index] > max_power) {
        max_power = bucket.size() * v[index];
        min_shrooms = bucket.size();
      }

      ++pick_index;
    }
  }

  cout << max_power << ' ' << min_shrooms << endl;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(0);

  ll t;
  cin >> t;
  while (t--) {
    solve();
  }
}

File: cf_935_div3/f.rs
use std::io;
use std::str::FromStr;
use std::collections::HashSet;

fn solve() {
  let n = read1::<usize>();
  // Read vecs starting from index 1.
  let mut v = read_vec_1::<us>(0);
  let p = read_vec_1::<us>(0);

  // Sort sequence of indexes 0..n according to descending of v[i].
  let mut indexes: Vec<us> = (0..=n).collect();
  indexes.sort_by(|&i, &j| v[j].partial_cmp(&v[i]).unwrap());

  // Keep indexes of shrooms which we picked.
  let mut hs: HashSet<us> = HashSet::new();

  // Next shroom to pick.
  let mut pick_index = 0;

  let mut max_power = 0;
  let mut min_shrooms = n;

  for take in 0..=n {
    // If we should consider permutation.
    if take >= 2 {
      let remove = p[take - 1];
      v[remove] = 0;
      hs.remove(&remove);
    }
    while hs.len() < take && pick_index <= n {
      // Picking shroom with index indexes[pick_index].
      let index = indexes[pick_index];
      // If it is valid.
      if v[index] > 0 {
        hs.insert(index);
      }
      if hs.len() * v[index] > max_power {
        max_power = hs.len() * v[index];
        min_shrooms = hs.len();
      }
      pick_index += 1;
    }
  }

  println!("{max_power} {min_shrooms}");
}

fn main() {
  for _ in 0..read1() {
    solve();
  }
}

#[allow(non_camel_case_types)]
type us = usize;

fn read_vec_1<T: FromStr>(first: T) -> Vec<T> {
  let mut v = read_vec::<T>();
  v.insert(0, first);
  v
}

fn read_vec<T: FromStr>() -> Vec<T> {
  read_line().split_whitespace().map(|x| x.parse().ok().unwrap()).collect::<Vec<T>>()
}

fn read_line() -> String {
  let mut input = String::new();
  io::stdin().read_line(&mut input).unwrap();
  input
}

fn read1<T: FromStr>() -> T {
  read_line().trim().parse().ok().unwrap()
}

fn read2<T1: FromStr, T2: FromStr>() -> (T1, T2) {
  let input = read_line();
  let mut i = input.split_whitespace();
  let a: T1 = i.next().unwrap().parse().ok().unwrap();
  let b: T2 = i.next().unwrap().parse().ok().unwrap();
  (a, b)
}
